pycobertura/reporters.py

```diff
@@ -93,152 +70,147 @@ def __init__(self, *args, **kwargs):
         self.differ = CoberturaDiff(cobertura1, cobertura2)
         self.show_source = show_source
+        self.color = kwargs.pop("color", False)
 
-    def get_file_row(self, filename):
-        row_values = [
-            filename,
-            self.differ.diff_total_statements(filename),
-            self.differ.diff_total_misses(filename),
-            self.differ.diff_line_rate(filename),
-        ]
+    @staticmethod
+    def format_line_rate(line_rate):
+        return f"{line_rate:+.2%}" if line_rate else "-"
 
-        if self.show_source is True:
-            missed_lines = self.differ.diff_missed_lines(filename)
-            row_values.append(missed_lines)
-            row = file_row_missed(*row_values)
-        else:
-            row = file_row(*row_values)
+    @staticmethod
+    def format_total_statements(total_statements):
+        return f"{total_statements:+d}" if total_statements else "-"
 
-        return row
+    @staticmethod
+    def format_missed_lines(missed_lines):
+        return [f"+{lno:d}" if is_new else f"-{lno:d}" for lno, is_new in missed_lines]
 
-    def get_footer_row(self):
-        footer_values = [
-            "TOTAL",
-            self.differ.diff_total_statements(),
-            self.differ.diff_total_misses(),
-            self.differ.diff_line_rate(),
-        ]
-
-        if self.show_source:
-            footer_values.append([])  # dummy missed lines
-            footer = file_row_missed(*footer_values)
-        else:
-            footer = file_row(*footer_values)
+    def color_row(self, row):
+        if self.color is True:
+            return red(row) if row[0] == "+" else green(row)
+        return row
 
-        return footer
+    def format_total_misses(self, total_misses):
+        return self.color_row(f"{total_misses:+d}") if total_misses else "-"
 
     def get_report_lines(self):
-        lines = []
-
-        for filename in self.differ.files():
-            file_row = self.get_file_row(filename)
-            if any(file_row[1:]):  # don't report unchanged class files
-                lines.append(file_row)
-
-        footer = self.get_footer_row()
-        lines.append(footer)
+        if not self.show_source:
+            row_values = tuple(
+                (
+                    filename,
+                    self.format_total_statements(
+                        self.differ.diff_total_statements(filename)
+                    ),
+                    self.format_total_misses(self.differ.diff_total_misses(filename)),
+                    self.format_line_rate(self.differ.diff_line_rate(filename)),
+                )
+                for filename in self.differ.files()
+                if any(
+                    (
+                        self.differ.diff_total_statements(filename),
+                        self.differ.diff_total_misses(filename),
+                        self.differ.diff_line_rate(filename),
```

Comment:
These are already calculated in the tuple comprehension above. Is there a way we compute them once and reuse them? It might require us to break down this "one-liner" comprehension.

Reply:
maybe...

Reply:
I could do a list comprehension for each and then loop through the 3 lists to evaluate the `filenames_of_files_with_changes` - I am just not sure we gain much, but I can try to do if you want me to

Reply:
addressed in commit 503ec5e